* __Step 1: Importing the Relevant Libraries__

* __Step 2: Reading Data__
    
* __Step 3: Data Inspection__
    
* __Step 4: Exploratory Data Analysis__
    
* __Step 5: Data Cleaning__
    
* __Step 6: Feature Engineering__

* __Step 7: Building Model__

###  Step 1: Importing the Relevant Libraries

In [ ]:
# Basic eda library
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Pre-Processing Library 
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

# Imputer 
from sklearn.impute import KNNImputer

#
from sklearn.model_selection import train_test_split

# Metrix
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

# To hide or filtering Warning for clear view
import warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

#### Customising Visulization

In [ ]:
sns.set(rc = {'figure.figsize':(16,5)})
sns.set_style('whitegrid')

## Step 2: Reading Data

In [ ]:
train_df = pd.read_csv('/content/drive/MyDrive/Dataset/InsurancePolicy/train_Df64byy.csv')
test_df = pd.read_csv('/content/drive/MyDrive/Dataset/InsurancePolicy/test_YCcRUnU.csv')

In [ ]:
train_df.head()

ID City_Code  Region_Code  ... Reco_Policy_Cat Reco_Policy_Premium  Response
0   1        C3         3213  ...              22             11628.0         0
1   2        C5         1117  ...              22             30510.0         0
2   3        C5         3732  ...              19              7450.0         1
3   4       C24         4378  ...              19             17780.0         0
4   5        C8         2190  ...              16             10404.0         0

[5 rows x 14 columns]

In [ ]:
test_df.head()

ID City_Code  ...  Reco_Policy_Cat Reco_Policy_Premium
0  50883        C1  ...                5             11934.0
1  50884        C4  ...               18             32204.8
2  50885        C1  ...               17              9240.0
3  50886        C3  ...               18              9086.0
4  50887        C1  ...                5             22534.0

[5 rows x 13 columns]

## Step 3: Data Inspection

In [ ]:
train_df.shape, test_df.shape

((50882, 14), (21805, 13))

* __We have 50882 rows and 14 columns in Train set whereas Test set has 21805 rows and 13 columns.__

In [ ]:
# Sum of null values in train data
train_df.isnull().sum()

ID                             0
City_Code                      0
Region_Code                    0
Accomodation_Type              0
Reco_Insurance_Type            0
Upper_Age                      0
Lower_Age                      0
Is_Spouse                      0
Health Indicator           11691
Holding_Policy_Duration    20251
Holding_Policy_Type        20251
Reco_Policy_Cat                0
Reco_Policy_Premium            0
Response                       0
dtype: int64

In [ ]:
# Sum of null values in test data
test_df.isnull().sum()

ID                            0
City_Code                     0
Region_Code                   0
Accomodation_Type             0
Reco_Insurance_Type           0
Upper_Age                     0
Lower_Age                     0
Is_Spouse                     0
Health Indicator           5027
Holding_Policy_Duration    8603
Holding_Policy_Type        8603
Reco_Policy_Cat               0
Reco_Policy_Premium           0
dtype: int64

In [ ]:
# Categorical Data
train_categorical = train_df.select_dtypes(include=[np.object])
obj_col = [col for col in train_df.keys() if train_df[col].dtype=='O']
print('Categorical Data of train data -->', train_categorical.shape[1],'\n',  obj_col)
print('*****************************************************************************')

# Numerical Data
train_numerical = train_df.select_dtypes(include=[np.int64])
print('Numerical Data of train data -->', train_numerical.shape[1])

Categorical Data of train data --> 6 
 ['City_Code', 'Accomodation_Type', 'Reco_Insurance_Type', 'Is_Spouse', 'Health Indicator', 'Holding_Policy_Duration']
*****************************************************************************
Numerical Data of train data --> 6


In [ ]:
# Categorical Data
test_categorical = test_df.select_dtypes(include=[np.object])
obj_col_test = [col for col in test_df.keys() if test_df[col].dtype=='O']
print('Categorical Data of test data -->', test_categorical.shape[1],'\n',  obj_col_test)
print('*****************************************************************************')

# Numerical Data
test_numerical = test_df.select_dtypes(include=[np.int64])
print('Numerical Data of test data -->', test_numerical.shape[1])

Categorical Data of test data --> 6 
 ['City_Code', 'Accomodation_Type', 'Reco_Insurance_Type', 'Is_Spouse', 'Health Indicator', 'Holding_Policy_Duration']
*****************************************************************************
Numerical Data of test data --> 5


## Step 5: Data Cleaning

**Data Cleaning required in these column**
- Health Indicator        --->  __11691__
- Holding_Policy_Duration --->  __20251__
- Holding_Policy_Type     --->  __20251__

* __5.1 Lets start with Health Indicator column.__

In [ ]:
train_df['Health Indicator'].unique()

array(['X1', 'X2', nan, 'X4', 'X3', 'X6', 'X5', 'X8', 'X7', 'X9'],
      dtype=object)

In [ ]:
def imp(col):
    train_df['Health Indicator']
  
    if col=='X1':
        return 1
    elif col=='X2':
        return 2
    elif col=='X3':
        return 3
    elif col=='X4':
        return 4
    elif col=='X5':
        return 5
    elif col=='X6':
        return 6
    elif col=='X7':
        return 7
    elif col=='X8':
        return 8
    elif col=='X9':
        return 9
    else:
        return np.nan

In [ ]:
#train Data 
train_df['Health Indicator'] = train_df['Health Indicator'].apply(imp).astype(float)
train_df['Health Indicator'] = train_df['Health Indicator'].replace('nan',np.nan)
# Test Data
test_df['Health Indicator'] = test_df['Health Indicator'].apply(imp).astype(float)
test_df['Health Indicator'] = test_df['Health Indicator'].replace('nan',np.nan)
train_df['Health Indicator'].dtypes

dtype('float64')

In [ ]:
train_df['Health Indicator'].unique()

array([ 1.,  2., nan,  4.,  3.,  6.,  5.,  8.,  7.,  9.])

In [ ]:
test_df['Health Indicator'].unique()

array([nan,  1.,  3.,  2.,  5.,  4.,  7.,  6.,  9.,  8.])

* __5.2 Holding_Policy_Duration column.__

In [ ]:
train_df['Holding_Policy_Duration'].unique()

array(['14+', nan, '1.0', '3.0', '5.0', '9.0', '14.0', '7.0', '2.0',
       '11.0', '10.0', '8.0', '6.0', '4.0', '13.0', '12.0'], dtype=object)

In [ ]:
test_df['Holding_Policy_Duration'].unique()

array(['6.0', '3.0', '2.0', nan, '14+', '5.0', '1.0', '4.0', '12.0',
       '11.0', '7.0', '9.0', '13.0', '8.0', '14.0', '10.0'], dtype=object)

In [ ]:

train_df['Holding_Policy_Duration'] = train_df['Holding_Policy_Duration'].replace('14+','14')

test_df['Holding_Policy_Duration'] = test_df['Holding_Policy_Duration'].replace('14+','14')

In [ ]:
test_df['Holding_Policy_Duration'].unique()

array(['6.0', '3.0', '2.0', nan, '14', '5.0', '1.0', '4.0', '12.0',
       '11.0', '7.0', '9.0', '13.0', '8.0', '14.0', '10.0'], dtype=object)

In [ ]:
train_df['Holding_Policy_Duration'].unique()

array(['14', nan, '1.0', '3.0', '5.0', '9.0', '14.0', '7.0', '2.0',
       '11.0', '10.0', '8.0', '6.0', '4.0', '13.0', '12.0'], dtype=object)

* __5.3 Holding_Policy_Type column.__

In [ ]:
train_df['Holding_Policy_Type'].unique()

array([ 3., nan,  1.,  4.,  2.])

In [ ]:
test_df['Holding_Policy_Type'].unique()

array([ 3.,  4., nan,  1.,  2.])

In [ ]:
train_df.head()

ID City_Code  Region_Code  ... Reco_Policy_Cat Reco_Policy_Premium  Response
0   1        C3         3213  ...              22             11628.0         0
1   2        C5         1117  ...              22             30510.0         0
2   3        C5         3732  ...              19              7450.0         1
3   4       C24         4378  ...              19             17780.0         0
4   5        C8         2190  ...              16             10404.0         0

[5 rows x 14 columns]

## __5.5 Label Encoder.__

In [ ]:
# Intiallising Encoder
encoder = LabelEncoder()

In [ ]:
train_df['City_Code'] = encoder.fit_transform(train_df['City_Code'])
test_df['City_Code'] = encoder.fit_transform(test_df['City_Code'])

In [ ]:
train_df['Accomodation_Type'] = pd.get_dummies(train_df['Accomodation_Type'],drop_first=False)
train_df['Reco_Insurance_Type'] = pd.get_dummies(train_df['Reco_Insurance_Type'],drop_first=False)
train_df['Is_Spouse'] = pd.get_dummies(train_df['Is_Spouse'],drop_first=False)

In [ ]:
test_df['Accomodation_Type'] = pd.get_dummies(test_df['Accomodation_Type'],drop_first=False)
test_df['Reco_Insurance_Type'] = pd.get_dummies(test_df['Reco_Insurance_Type'],drop_first=False)
test_df['Is_Spouse'] = pd.get_dummies(test_df['Is_Spouse'],drop_first=False)

In [ ]:
len(test_df['Health Indicator'].unique())

10

In [ ]:
len(test_df['Reco_Insurance_Type'].unique())

2

In [ ]:
test_df['Reco_Insurance_Type'].value_counts()

1    17341
0     4464
Name: Reco_Insurance_Type, dtype: int64

In [ ]:
len(test_df['Accomodation_Type'].unique())

2

In [ ]:
test_df['Accomodation_Type'].value_counts()

1    12090
0     9715
Name: Accomodation_Type, dtype: int64

1    12090
0     9715

## __5.6 Missing Value Treatment.__

### __KNN Imputer .__

In [ ]:
knnimputer = KNNImputer(n_neighbors=5)

In [ ]:
key_tr = train_df.keys()
key_ts = test_df.keys()

In [ ]:
train_df = knnimputer.fit_transform(train_df)
test_df = knnimputer.fit_transform(test_df)

In [ ]:
train_df=pd.DataFrame(train_df,columns=key_tr)
test_df=pd.DataFrame(test_df,columns=key_ts)

In [ ]:
train_df

ID  City_Code  ...  Reco_Policy_Premium  Response
0          1.0       22.0  ...              11628.0       0.0
1          2.0       31.0  ...              30510.0       0.0
2          3.0       31.0  ...               7450.0       1.0
3          4.0       16.0  ...              17780.0       0.0
4          5.0       34.0  ...              10404.0       0.0
...        ...        ...  ...                  ...       ...
50877  50878.0       30.0  ...               7704.0       0.0
50878  50879.0       31.0  ...               5408.0       0.0
50879  50880.0        0.0  ...              11374.0       0.0
50880  50881.0        0.0  ...              28179.2       1.0
50881  50882.0       22.0  ...              11424.0       0.0

[50882 rows x 14 columns]

In [ ]:
len(train_df['Health Indicator'].unique())

27

In [ ]:
train_df['Health Indicator'].unique()

array([1. , 2. , 3.2, 2.8, 4. , 3. , 2.6, 6. , 5. , 1.4, 1.8, 3.4, 2.2,
       4.8, 1.6, 2.4, 3.6, 1.2, 3.8, 8. , 7. , 4.2, 4.6, 4.4, 9. , 5.2,
       5.8])

In [ ]:
np.ceil(train_df['Health Indicator'])

0        1.0
1        2.0
2        4.0
3        1.0
4        2.0
        ... 
50877    3.0
50878    3.0
50879    2.0
50880    2.0
50881    3.0
Name: Health Indicator, Length: 50882, dtype: float64

In [ ]:
len(test_df['Reco_Insurance_Type'].unique())

2

In [ ]:
len(test_df['Accomodation_Type'].unique())

2

In [ ]:
train_df.isnull().sum()

ID                         0
City_Code                  0
Region_Code                0
Accomodation_Type          0
Reco_Insurance_Type        0
Upper_Age                  0
Lower_Age                  0
Is_Spouse                  0
Health Indicator           0
Holding_Policy_Duration    0
Holding_Policy_Type        0
Reco_Policy_Cat            0
Reco_Policy_Premium        0
Response                   0
dtype: int64

In [ ]:
test_df.isnull().sum()

ID                         0
City_Code                  0
Region_Code                0
Accomodation_Type          0
Reco_Insurance_Type        0
Upper_Age                  0
Lower_Age                  0
Is_Spouse                  0
Health Indicator           0
Holding_Policy_Duration    0
Holding_Policy_Type        0
Reco_Policy_Cat            0
Reco_Policy_Premium        0
dtype: int64

In [ ]:
X=train_df.drop(['Response', 'ID', 'City_Code', 'Region_Code'],axis=1)
y=train_df[['Response']]

In [ ]:
X

Accomodation_Type  ...  Reco_Policy_Premium
0                    0.0  ...              11628.0
1                    1.0  ...              30510.0
2                    1.0  ...               7450.0
3                    1.0  ...              17780.0
4                    0.0  ...              10404.0
...                  ...  ...                  ...
50877                0.0  ...               7704.0
50878                0.0  ...               5408.0
50879                0.0  ...              11374.0
50880                1.0  ...              28179.2
50881                0.0  ...              11424.0

[50882 rows x 10 columns]

In [ ]:
y

Response
0           0.0
1           0.0
2           1.0
3           0.0
4           0.0
...         ...
50877       0.0
50878       0.0
50879       0.0
50880       1.0
50881       0.0

[50882 rows x 1 columns]

In [ ]:
Xtr = X.columns

In [ ]:
test_df

ID  City_Code  ...  Reco_Policy_Cat  Reco_Policy_Premium
0      50883.0        0.0  ...              5.0              11934.0
1      50884.0       30.0  ...             18.0              32204.8
2      50885.0        0.0  ...             17.0               9240.0
3      50886.0       22.0  ...             18.0               9086.0
4      50887.0        0.0  ...              5.0              22534.0
...        ...        ...  ...              ...                  ...
21800  72683.0       22.0  ...             18.0              15884.0
21801  72684.0       30.0  ...             18.0              21390.0
21802  72685.0        3.0  ...              1.0              17836.0
21803  72686.0        1.0  ...             19.0              11568.0
21804  72687.0       11.0  ...              3.0              23040.0

[21805 rows x 13 columns]

In [ ]:
test_df.head(1)

ID  City_Code  ...  Reco_Policy_Cat  Reco_Policy_Premium
0  50883.0        0.0  ...              5.0              11934.0

[1 rows x 13 columns]

In [ ]:
X.head(1)

Accomodation_Type  Reco_Insurance_Type  ...  Reco_Policy_Cat  Reco_Policy_Premium
0                0.0                  1.0  ...             22.0              11628.0

[1 rows x 10 columns]

## Standard Scalling

In [ ]:
scaler = StandardScaler()

In [ ]:
X = scaler.fit_transform(X)
test_df = scaler.fit_transform(test_df)

In [ ]:
X.shape

(50882, 10)

In [ ]:
X=pd.DataFrame(X,columns=Xtr)
# y=pd.DataFrame(y,columns=train_df.Response)
test_df=pd.DataFrame(test_df,columns=key_ts)

In [ ]:
y

Response
0           0.0
1           0.0
2           1.0
3           0.0
4           0.0
...         ...
50877       0.0
50878       0.0
50879       0.0
50880       1.0
50881       0.0

[50882 rows x 1 columns]

In [ ]:
## Train Test Split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state =101)

##Implementing DecisionTree

In [ ]:
from sklearn.tree import DecisionTreeClassifier
DT = DecisionTreeClassifier('gini',max_depth=3)
DT.fit(X_train,y_train)
DT_pred = DT.predict(X_test)
DT_pred

array([0., 0., 0., ..., 0., 0., 0.])

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,DT_pred)

0.7642974123812644

In [ ]:
DT.score(X_test,y_test)

0.7642974123812644

In [ ]:
DT.score(X_train,y_train)

0.7582334278574838

##Implementing Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=100,max_depth=10,criterion='entropy')
rf.fit(X_train,y_train)
rf_pred = rf.predict(X_test)
accuracy_score(y_test,rf_pred)

0.7642974123812644

In [ ]:
rf.score(X_test,y_test)

0.7642974123812644

In [ ]:
rf.score(X_train,y_train)

0.7589914928264593

##Implementing XGBoost

In [ ]:
from xgboost import XGBClassifier
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV


In [ ]:
## Hyper Parameter Optimization

params={
 "learning_rate"    : [0.05, 0.10, 0.15, 0.20, 0.25, 0.30 ] ,
 "max_depth"        : [ 3, 4, 5, 6, 8, 10, 12, 15],
 "min_child_weight" : [ 1, 3, 5, 7 ],
 "gamma"            : [ 0.0, 0.1, 0.2 , 0.3, 0.4 ],
 "colsample_bytree" : [ 0.3, 0.4, 0.5 , 0.7 ]
}

In [ ]:
xgboost_model = XGBClassifier(max_depth=10)

In [ ]:
random_search=RandomizedSearchCV(xgboost_model,param_distributions=params,n_iter=5,scoring='roc_auc',n_jobs=-1,cv=5)


In [ ]:
random_search.fit(X_train,y_train)

RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                           colsample_bylevel=1,
                                           colsample_bynode=1,
                                           colsample_bytree=1, gamma=0,
                                           learning_rate=0.1, max_delta_step=0,
                                           max_depth=10, min_child_weight=1,
                                           missing=None, n_estimators=100,
                                           n_jobs=1, nthread=None,
                                           objective='binary:logistic',
                                           random_state=0, reg_alpha=0,
                                           reg_lambda=1, s...
                                           verbosity=1),
                   iid='deprecated', n_iter=5, n_jobs=-1,
                   param_distributions={'colsample_bytree': [0

In [ ]:
random_search.best_estimator_

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5, gamma=0.1,
              learning_rate=0.1, max_delta_step=0, max_depth=12,
              min_child_weight=3, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
random_search.best_params_

{'colsample_bytree': 0.5,
 'gamma': 0.1,
 'learning_rate': 0.1,
 'max_depth': 12,
 'min_child_weight': 3}

In [ ]:
xgb = XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.7, gamma=0.2,
              learning_rate=0.15, max_delta_step=0, max_depth=5,
              min_child_weight=5, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
xgb.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.7, gamma=0.2,
              learning_rate=0.15, max_delta_step=0, max_depth=5,
              min_child_weight=5, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
xgb_pred = xgb.predict(X_test)

In [ ]:
xgb_pred

array([0., 0., 0., ..., 0., 0., 0.])

In [ ]:
score = accuracy_score(y_test,xgb_pred)
score

0.7637078283655421

In [ ]:
final_predictionht = xgb.predict(test_df.drop(['ID', 'City_Code', 'Region_Code'],axis=1))

In [ ]:
final_predictionht.shape

(21805,)

In [ ]:
a = pd.DataFrame(final_predictionht)

In [ ]:
a[0].unique()

array([0., 1.])

In [ ]:
a.value_counts()

0.0    21685
1.0      120
dtype: int64

In [ ]:
submission = pd.read_csv('/content/drive/MyDrive/Dataset/InsurancePolicy/sample_submission_QrCyCoT.csv')  # uploading submission file
final_predictions = xgbclassifier.predict(test_df.drop(['ID', 'City_Code', 'Region_Code'],axis=1))
submission['Response'] = final_predictions
#only positive predictions for the target variable

submission.to_csv('my_submission_xgboost.csv', index=False)

In [ ]:
my = pd.read_csv('my_submission_xgboost.csv')

In [ ]:
my['Response'].value_counts()

0.0    21646
1.0      159
Name: Response, dtype: int64